fdbserver/workloads/WriteTagThrottling.actor.cpp

```diff
@@ -0,0 +1,271 @@
+	                                      WriteTagThrottlingWorkload* self) {
+		state double lastTime = now();
+		state double opStart;
+		state StringRef key;
+		try {
+			loop {
+				wait(poisson(&lastTime, self->txInterval));
+				state double txStart;
+				state Transaction tx(cx);
+				state int i;
+				// give tag to client
+				if (self->writeThrottle) {
+					if (isBadActor) {
+						tx.options.tags.addTag(self->badWriteTag);
+						tx.options.tags.addTag(self->badReadTag);
+					} else if (deterministicRandom()->coinflip()) {
+						tx.options.tags.addTag(self->goodTag);
+					}
+				}
+				txStart = now();
+				while (true) {
+					try {
+						for (i = 0; i < self->numClearPerTx; ++i) {
+							bool useClearKey = deterministicRandom()->random01() < badOpRate;
+							tx.clear(self->generateRange(useClearKey, actorId));
+						}
+						for (i = 0; i < self->numWritePerTx; ++i) {
+							bool useReadKey = deterministicRandom()->random01() < badOpRate;
+							key = self->generateKey(useReadKey, actorId + self->keyCount / 3);
+							tx.set(key, self->generateVal());
+						}
+						for (i = 0; i < self->numReadPerTx; ++i) {
+							bool useReadKey = deterministicRandom()->random01() < badOpRate;
+							ASSERT(self->keyCount >= actorId);
+							key = self->generateKey(useReadKey, self->keyCount - actorId);
+							opStart = now();
+							Optional<Value> v = wait(tx.get(key));
+							double duration = now() - opStart;
+							isBadActor ? self->badActorReadLatency.addSample(duration)
+							           : self->goodActorReadLatency.addSample(duration);
+						}
+						opStart = now();
+						wait(tx.commit());
+						double duration = now() - opStart;
+						isBadActor ? self->badActorCommitLatency.addSample(duration)
+						           : self->goodActorCommitLatency.addSample(duration);
+						break;
+					} catch (Error& e) {
+						if (e.code() == error_code_transaction_too_old) {
+							isBadActor ? ++self->badActorTooOldRetries : ++self->goodActorTooOldRetries;
+						} else if (e.code() == error_code_not_committed) {
+							isBadActor ? ++self->badActorCommitFailedRetries : ++self->goodActorCommitFailedRetries;
+						} else if (e.code() == error_code_tag_throttled) {
+							isBadActor ? ++self->badActorThrottleRetries : ++self->goodActorThrottleRetries;
+						}
+						wait(tx.onError(e));
+					}
+					isBadActor ? ++self->badActorRetries : ++self->goodActorRetries;
+				}
+				double duration = now() - txStart;
+				if (isBadActor) {
+					++self->badActorTxNum;
+					self->badActorTotalLatency += duration;
+				} else {
+					++self->goodActorTxNum;
+					self->goodActorTotalLatency += duration;
+				}
+			}
+		} catch (Error& e) {
+			TraceEvent(SevError, "WriteThrottling").error(e);
+			throw;
+		}
+	}
+	// collect health metrics
+	//   ACTOR static Future<Void> healthMetricsChecker(Database cx, WriteTagThrottlingWorkload* self) {
```

Comment:
is this something todo?

Reply:
Yes. Basically, I would use this actor periodically to ask which tag is throttled and record it. Other health metrics also have chances to be added while the feature is being constructed.

Reply:
If it is useful, then add a comment to specify say "This is for debug purpose". Otherwise, another person may remove it accidentally because it is commented code.